lib/sample_rate_converter/sample_rate_converter.c

```diff
@@ -0,0 +1,406 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "sample_rate_converter.h"
+#include "sample_rate_converter_filter.h"
+
+#include <errno.h>
+#include <stdbool.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(sample_rate_converter, CONFIG_SAMPLE_RATE_CONVERTER_LOG_LEVEL);
+
+/**
+ * The input buffer must be able to store two samples in addition to the block size to meet
+ * filter requirements.
+ */
+#define INTERNAL_INPUT_BUF_NUMBER_SAMPLES                                                          \
+	(CONFIG_SAMPLE_RATE_CONVERTER_BLOCK_SIZE_MAX +                                             \
+	 SAMPLE_RATE_CONVERTER_INPUT_BUFFER_NUMBER_OVERFLOW_SAMPLES)
+
+#ifdef CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_16
+#define SAMPLE_RATE_CONVERTER_INTERNAL_INPUT_BUF_SIZE                                              \
+	(INTERNAL_INPUT_BUF_NUMBER_SAMPLES * sizeof(uint16_t))
+#define SAMLPE_RATE_CONVERTER_INTERNAL_OUTPUT_BUF_SIZE                                             \
+	(CONFIG_SAMPLE_RATE_CONVERTER_BLOCK_SIZE_MAX * sizeof(uint16_t))
+#elif CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_32
+#define SAMPLE_RATE_CONVERTER_INTERNAL_INPUT_BUF_SIZE                                              \
+	(INTERNAL_INPUT_BUF_NUMBER_SAMPLES * sizeof(uint32_t))
+#define SAMLPE_RATE_CONVERTER_INTERNAL_OUTPUT_BUF_SIZE                                             \
+	(CONFIG_SAMPLE_RATE_CONVERTER_BLOCK_SIZE_MAX * sizeof(uint32_t))
+#endif
+
+static int validate_sample_rates(uint32_t sample_rate_input, uint32_t sample_rate_output)
+{
+	if (sample_rate_input > sample_rate_output) {
+		if (sample_rate_input != 48000) {
+			LOG_ERR("Invalid input sample rate for downsampling %d", sample_rate_input);
+			return -EINVAL;
+		}
+
+		if ((sample_rate_output != 24000) && (sample_rate_output != 16000)) {
+			LOG_ERR("Invalid output sample rate for downsampling %d",
+				sample_rate_output);
+			return -EINVAL;
+		}
+	} else if (sample_rate_input < sample_rate_output) {
+		if (sample_rate_output != 48000) {
+			LOG_ERR("Invalid output sample rate for upsampling: %d",
+				sample_rate_output);
+			return -EINVAL;
+		}
+
+		if ((sample_rate_input != 24000) && (sample_rate_input != 16000)) {
+			LOG_ERR("Invalid input sample rate for upsampling: %d", sample_rate_input);
+			return -EINVAL;
+		}
+	} else {
+		LOG_ERR("Input and out sample rates are the same");
+		return -EINVAL;
+	}
+	return 0;
+}
+
+/**
+ * When upsampling from 16kHz to 48kHz the input and output bytes must be buffered. This is to
+ * fulfill the requirement for the filter that the number of input bytes must be divisible by the
+ * conversion factor. This function returns true when this is the case.
+ */
+static inline bool conversion_needs_buffering(struct sample_rate_converter_ctx *ctx,
+					      uint8_t conversion_ratio)
+{
+	return ((ctx->conversion_direction == CONVERSION_DIR_UP) && (conversion_ratio == 3));
```

Comment:
I like the idea, but I think this would complicate more cases than it simplifies. E.g where the code is currently checking for up/down it would have to check for two cases instead of one, and as the conversion ratio is needed as an integer several places it would have to be calculated either way.

Reply:
After offline discussion I will try to implement the ratio as a signed value